src/buffer/out/search.cpp

```diff
@@ -190,22 +191,26 @@ bool Search::_FindNeedleInHaystackAt(const til::point pos, til::point& start, ti
     start = {};
     end = {};
 
+    if (_needle.empty())
+    {
+        return false;
+    }
+
     auto bufferPos = pos;
 
-    for (const auto& needleCell : _needle)
+    for (const auto& str : til::utf16_iterator{ _needle })
     {
-        // Haystack is the buffer. Needle is the string we were given.
-        const auto hayIter = _uiaData.GetTextBuffer().GetTextDataAt(bufferPos);
-        const auto hayChars = *hayIter;
-        const auto needleChars = std::wstring_view(needleCell.data(), needleCell.size());
-
-        // If we didn't match at any point of the needle, return false.
-        if (!_CompareChars(hayChars, needleChars))
+        const auto width = IsGlyphFullWidth(str) ? 2 : 1;
+        for (int i = 0; i < width; ++i)

```

Comment:
If we get a search string of...

```
KATAKANA KA
```

and the buffer contains

```
KATAKANA KA
```

Do we iterate _twice_, because `KATAKANA KA` is 2 columns wide?

It looks like we do. `GetTextDataAt(0)` will return `KA` and then we'll increment by 1, and `GetTextDataAt(1)` will return `KA` again.

Is that on purpose?

Reply:
I guess: Search still pretends DBCS characters are stored "the old way," with two cells. Since you're already _here,_ shouldn't we fix that?

Reply:
I actually tried fixing it but it ended up being a significant rewrite and it didn't even work properly with wide characters, because `TextBufferCellIterator` isn't able to navigate over wide chars by itself either. `Renderer` does it manually and it's pretty complex code. 😒

if anything we should write a simple Boyer-Moore searcher, because we can then prepare the lookup table with precomputed lower-case and upper-case `char32_t` entries. That way you don't have to call `towlower` in a tight inner loop for every character.